Error when linking with libxdp

When linking with libxdp, two errors are generated:
```
cc -Wl,--no-as-needed -lxdp
/usr/bin/ld: /usr/local/lib/libxdp.so: undefined reference to `try_snprintf'
/usr/bin/ld: /usr/local/lib/libxdp.so: undefined reference to `silence_libbpf_logging'
```

objdump confirms they are undefined:
```
objdump -t /usr/local/lib/libxdp.so | grep try_snprintf
000000000000a710 l F .text 00000000000000da try_snprintf
0000000000000000 *UND* 0000000000000000 try_snprintf

objdump -t /usr/local/lib/libxdp.so | grep silence_libbpf
0000000000000000 *UND* 0000000000000000 silence_libbpf_logging
```

The call to try_snprintf was introduced to xsk.c in:
fdd1566dcd1f ("libxdp: get queue num from sysfs in xsk")

The call to silence_libbpf_logging in xsk.c was introduced in:
ea481a18ed07 ("libxdp: move AF_XDP functionality from libbpf")

Ciara Loftus ***@***.***> writes:

> When linking with libxdp, two errors are generated:
> ```
> cc -Wl,--no-as-needed -lxdp
> /usr/bin/ld: /usr/local/lib/libxdp.so: undefined reference to `try_snprintf'
> /usr/bin/ld: /usr/local/lib/libxdp.so: undefined reference to `silence_libbpf_logging'
> ```
>
> objdump confirms they are undefined:
> ```
> objdump -t /usr/local/lib/libxdp.so | grep try_snprintf
> 000000000000a710 l F .text 00000000000000da try_snprintf
> 0000000000000000 *UND* 0000000000000000 try_snprintf
>
> objdump -t /usr/local/lib/libxdp.so | grep silence_libbpf
> 0000000000000000 *UND* 0000000000000000 silence_libbpf_logging
> ```
>
> The call to try_snprintf was introduced to xsk.c in:
> fdd1566dcd1f ("libxdp: get queue num from sysfs in xsk")
>
> The call to silence_libbpf_logging in xsk.c was introduced in:
> ea481a18ed07 ("libxdp: move AF_XDP functionality from libbpf")

Oops, that's embarrassing! Those definitely shouldn't be there at all...

--- a/Makefile
+++ b/Makefile
@@ -72,7 +72,7 @@ install: all
 	do $(MAKE) -C $$i install; done
 
 test: all
-	@for i in $(UTILS); do \
+	@for i in lib/libxdp $(UTILS); do \
 		echo; echo test $$i; $(MAKE) -C $$i test; \
 		if [ $$? -ne 0 ]; then failed="y"; fi; \
 	done; \

--- a/headers/xdp/libxdp.h
+++ b/headers/xdp/libxdp.h
@@ -11,6 +11,7 @@
 
 #include <stdio.h>
 #include <linux/bpf.h>
+#include <bpf/libbpf.h>
 #include <bpf/bpf.h>
 #include "xdp_helpers.h"
 

--- a/lib/libxdp/Makefile
+++ b/lib/libxdp/Makefile
@@ -16,6 +16,10 @@ STATIC_OBJS := $(addprefix $(STATIC_OBJDIR)/,$(OBJS))
 STATIC_LIBS := $(OBJDIR)/libxdp.a
 MAN_PAGE := libxdp.3
 MAN_OBJ := ${MAN_PAGE:.3=.man}
+TEST_DIR := tests
+TEST_FILE := $(TEST_DIR)/test-libxdp.sh
+TEST_CFLAGS := $(CFLAGS) -I$(realpath $(HEADER_DIR)) -L$(realpath $(OBJDIR)) -Wall -Werror $(LDFLAGS)
+TEST_LDLIBS := $(LDLIBS)
 
 SHARED_CFLAGS += -fPIC -DSHARED
 LIB_HEADERS := $(wildcard $(HEADER_DIR)/xdp/*.h)
@@ -153,3 +157,7 @@ $(MAN_PAGE): $(MAN_OBJ)
 	$(QUIET_GEN)sed -e "1 s/DATE/$(shell date '+%B %_d, %Y')/" -e "1 s/VERSION/v$(TOOLS_VERSION)/" $< > $@
 
 endif
+
+.PHONY: test
+test: all
+	$(Q)env CC="$(CC)" CFLAGS="$(TEST_CFLAGS)" LDLIBS="$(TEST_LDLIBS)" V=$(V) $(TEST_DIR)/test_runner.sh $(TEST_FILE)

--- a/lib/libxdp/Makefile
+++ b/lib/libxdp/Makefile
@@ -4,7 +4,6 @@ LIB_DIR = ..
 
 include libxdp.mk
 include $(LIB_DIR)/defines.mk
-include $(LIB_DIR)/util/util.mk
 
 OBJDIR ?= .
 SHARED_OBJDIR := $(OBJDIR)/sharedobjs
@@ -28,7 +27,7 @@ EXTRA_LIB_DEPS := $(OBJECT_LIBBPF) $(LIBMK) $(LIB_OBJS) $(LIB_HEADERS) compat.h 
 PC_FILE := $(OBJDIR)/libxdp.pc
 TEMPLATED_SOURCES := xdp-dispatcher.c
 
-CFLAGS += -I$(HEADER_DIR) -I$(LIB_DIR)/util
+CFLAGS += -I$(HEADER_DIR)
 BPF_CFLAGS += -I$(HEADER_DIR)
 
 

--- a/lib/libxdp/libxdp.c
+++ b/lib/libxdp/libxdp.c
@@ -152,23 +152,6 @@ static char *libxdp_strerror_r(int err, char *dst, size_t size)
 	return dst;
 }
 
-static int try_snprintf(char *buf, size_t buf_len, const char *format, ...)
-{
-	va_list args;
-	int len;
-
-	va_start(args, format);
-	len = vsnprintf(buf, buf_len, format, args);
-	va_end(args);
-
-	if (len < 0)
-		return -EINVAL;
-	else if ((size_t)len >= buf_len)
-		return -ENAMETOOLONG;
-
-	return 0;
-}
-
 static bool bpf_is_valid_mntpt(const char *mnt, unsigned long magic)
 {
 	struct statfs st_fs;

--- a/lib/libxdp/libxdp_internal.h
+++ b/lib/libxdp/libxdp_internal.h
@@ -1,13 +1,33 @@
 #ifndef __LIBXDP_LIBXDP_INTERNAL_H
 #define __LIBXDP_LIBXDP_INTERNAL_H
 
+#include <unistd.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <stdarg.h>
+#include <stdio.h>
 #include <xdp/libxdp.h>
 
 #define LIBXDP_HIDE_SYMBOL __attribute__((visibility("hidden")))
 
 #define __printf(a, b) __attribute__((format(printf, a, b)))
 
-extern void silence_libbpf_logging(void);
+static inline int try_snprintf(char *buf, size_t buf_len, const char *format, ...)
+{
+	va_list args;
+	int len;
+
+	va_start(args, format);
+	len = vsnprintf(buf, buf_len, format, args);
+	va_end(args);
+
+	if (len < 0)
+		return -EINVAL;
+	else if ((size_t)len >= buf_len)
+		return -ENAMETOOLONG;
+
+	return 0;
+}
 
 LIBXDP_HIDE_SYMBOL __printf(2, 3) void libxdp_print(enum libxdp_print_level level,
 						    const char *format, ...);
@@ -23,5 +43,22 @@ LIBXDP_HIDE_SYMBOL __printf(2, 3) void libxdp_print(enum libxdp_print_level leve
 LIBXDP_HIDE_SYMBOL int check_xdp_prog_version(const struct btf *btf, const char *name,
 					      __u32 *version);
 LIBXDP_HIDE_SYMBOL struct xdp_program *xdp_program__clone(struct xdp_program *prog);
+
+#define min(x, y) ((x) < (y) ? x : y)
+#define max(x, y) ((x) > (y) ? x : y)
+
+#ifndef offsetof
+#define offsetof(type, member) ((size_t) & ((type *)0)->member)
+#endif
+
+#ifndef container_of
+#define container_of(ptr, type, member)                            \
+	({                                                         \
+		const typeof(((type *)0)->member) *__mptr = (ptr); \
+		(type *)((char *)__mptr - offsetof(type, member)); \
+	})
+#endif
+
+#define ARRAY_SIZE(x) (sizeof(x) / sizeof(*(x)))
 
 #endif /* __LIBXDP_LIBXDP_INTERNAL_H */

--- a/lib/libxdp/tests/test-libxdp.sh
+++ b/lib/libxdp/tests/test-libxdp.sh
@@ -1,0 +1,27 @@
+ALL_TESTS="test_link_so test_link_a"
+
+test_link_so()
+{
+        cat >$TMPDIR/libxdptest.c <<EOF
+#include <xdp/libxdp.h>
+int main(int argc, char **argv) {
+    (void) argc; (void) argv;
+    (void) xdp_program__open_file("filename", "section_name", NULL);
+    return 0;
+}
+EOF
+        check_run $CC -o $TMPDIR/libxdptest $TMPDIR/libxdptest.c $CFLAGS -lxdp $LDLIBS 2>&1
+}
+
+test_link_a()
+{
+        cat >$TMPDIR/libxdptest.c <<EOF
+#include <xdp/libxdp.h>
+int main(int argc, char **argv) {
+    (void) argc; (void) argv;
+    (void) xdp_program__open_file("filename", "section_name", NULL);
+    return 0;
+}
+EOF
+        check_run $CC -o $TMPDIR/libxdptest $TMPDIR/libxdptest.c $CFLAGS -l:libxdp.a $LDLIBS 2>&1
+}

--- a/lib/libxdp/tests/test_runner.sh
+++ b/lib/libxdp/tests/test_runner.sh
@@ -1,0 +1,100 @@
+#!/bin/bash
+# SPDX-License-Identifier: GPL-2.0-or-later
+#
+# Script to setup and manage tests for xdp-tools.
+# Based on the test-env script from xdp-tutorial.
+#
+# Author:   Toke Høiland-Jørgensen (toke@redhat.com)
+# Date:     26 May 2020
+# Copyright (c) 2020 Red Hat
+
+set -o errexit
+set -o nounset
+umask 077
+
+TEST_PROG_DIR="${TEST_PROG_DIR:-$(dirname "${BASH_SOURCE[0]}")}"
+ALL_TESTS=""
+VERBOSE_TESTS=${V:-0}
+
+TMPDIR=$(mktemp --tmpdir -d config.XXXXXX)
+trap 'status=$?; rm -rf $TMPDIR; exit $status' EXIT HUP INT QUIT TERM
+
+
+# Odd return value for skipping, as only 0-255 is valid.
+SKIPPED_TEST=249
+
+is_func()
+{
+    type "$1" 2>/dev/null | grep -q 'is a function'
+}
+
+check_run()
+{
+    local ret
+
+    [ "$VERBOSE_TESTS" -eq "1" ] && echo "$@"
+    "$@"
+    ret=$?
+    if [ "$ret" -ne "0" ]; then
+        exit $ret
+    fi
+}
+
+exec_test()
+{
+    local testn="$1"
+    local output
+    local ret
+
+    printf "     %-30s" "[$testn]"
+    if ! is_func "$testn"; then
+        echo "INVALID"
+        return 1
+    fi
+
+    output=$($testn 2>&1)
+    ret=$?
+    if [ "$ret" -eq "0" ]; then
+        echo "PASS"
+    elif [ "$ret" -eq "$SKIPPED_TEST" ]; then
+        echo "SKIPPED"
+        ret=0
+    else
+        echo "FAIL"
+    fi
+    if [ "$ret" -ne "0" ] || [ "$VERBOSE_TESTS" -eq "1" ]; then
+        echo "$output" | sed 's/^/\t/'
+    fi
+    return $ret
+}
+
+run_tests()
+{
+    local TESTS="$*"
+    local ret=0
+    [ -z "$TESTS" ] && TESTS="$ALL_TESTS"
+
+    echo "    Running tests from $TEST_DEFINITIONS"
+
+    for testn in $TESTS; do
+        exec_test $testn || ret=1
+        if is_func cleanup_tests; then
+            cleanup_tests || true
+        fi
+    done
+
+    return $ret
+}
+
+usage()
+{
+    echo "Usage: $0 <test_definition_file> [test names]" >&2
+    exit 1
+}
+
+TEST_DEFINITIONS="${1:-}"
+[ -f "$TEST_DEFINITIONS" ] || usage
+source "$TEST_DEFINITIONS"
+
+shift
+run_tests "$@"

--- a/lib/libxdp/xsk.c
+++ b/lib/libxdp/xsk.c
@@ -33,7 +33,6 @@
 #include <xdp/xsk.h>
 
 #include "libxdp_internal.h"
-#include "util.h"
 #include "xsk_def_xdp_prog.h"
 #include "bpf_instr.h"
 
@@ -638,8 +637,6 @@ static int __xsk_setup_xdp_prog(struct xsk_socket *xsk, int *xsks_map_fd)
 	struct xdp_program *prog;
 	const char *file_name;
 	int err;
-
-	silence_libbpf_logging();
 
 	prog = xsk_lookup_program(ctx->ifindex);
 	if (IS_ERR(prog))

